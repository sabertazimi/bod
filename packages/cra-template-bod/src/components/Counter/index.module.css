.row {
  @apply mb-4 flex items-center justify-center;
}

.row > button {
  @apply ml-1 mr-2;
}

.value {
  @apply mt-0.5 px-4 text-8xl;

  font-family: 'Courier New', Courier, monospace;
}

.textbox {
  @apply mr-1 w-16 border border-primary p-1 text-center text-4xl;
}

.button,
.async-button {
  @apply px-3 pb-1 text-4xl;

  color: rgb(112 76 182);
  appearance: none;
  cursor: pointer;
  background: none;
  background-color: rgb(112 76 182 / 10%);
  border: 2px solid transparent;
  border-radius: 2px;
  outline: none;
}

.async-button {
  position: relative;
}

@media (prefers-reduced-motion: no-preference) {
  .button,
  .async-button {
    transition: all 0.15s;
  }
}

.button:hover,
.button:focus,
.async-button:hover,
.async-button:focus {
  border: 2px solid rgb(112 76 182 / 40%);
}

.button:active,
.async-button:active {
  background-color: rgb(112 76 182 / 20%);
}

.async-button::after {
  position: absolute;
  top: 0;
  left: 0;
  display: block;
  width: 100%;
  height: 100%;
  content: '';
  background-color: rgb(112 76 182 / 15%);
  opacity: 0;
}

.async-button:active::after {
  width: 0%;
  opacity: 1;
}

@media (prefers-reduced-motion: no-preference) {
  .async-button::after {
    transition:
      width 1s linear,
      opacity 0.5s ease 1s;
  }

  .async-button:active::after {
    transition: 0s;
  }
}
