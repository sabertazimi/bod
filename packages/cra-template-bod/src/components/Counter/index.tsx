import { AppDispatch } from '../../store'
import styles from './index.module.css'

interface CounterProps {
  count: number
  incrementAmount: string
  onIncrementAmountChange: (value: string) => void
  onDecrement: () => ReturnType<AppDispatch>
  onIncrement: () => ReturnType<AppDispatch>
  onIncrementAsync: () => void
  onIncrementByAmount: () => ReturnType<AppDispatch>
  onIncrementIfOdd: () => void
}

/**
 * Counter
 * @param {CounterProps} Count props
 * @returns {JSX.Element} Counter component
 */
function Counter({
  count,
  incrementAmount,
  onIncrementAmountChange,
  onDecrement,
  onIncrement,
  onIncrementAsync,
  onIncrementByAmount,
  onIncrementIfOdd,
}: CounterProps): JSX.Element {
  return (
    <div>
      <div className={styles.row}>
        <button
          className={styles.button}
          aria-label="Decrement value"
          onClick={onDecrement}
        >
          -
        </button>
        <span className={styles.value}>{count}</span>
        <button
          className={styles.button}
          aria-label="Increment value"
          onClick={onIncrement}
        >
          +
        </button>
      </div>
      <div className={styles.row}>
        <input
          className={styles.textbox}
          aria-label="Set increment amount"
          value={incrementAmount}
          onChange={e => onIncrementAmountChange(e.target.value)}
        />
        <button className={styles.button} onClick={onIncrementByAmount}>
          Add Amount
        </button>
        <button className={styles.asyncButton} onClick={onIncrementAsync}>
          Add Async
        </button>
        <button className={styles.button} onClick={onIncrementIfOdd}>
          Add If Odd
        </button>
      </div>
    </div>
  )
}

export default Counter
