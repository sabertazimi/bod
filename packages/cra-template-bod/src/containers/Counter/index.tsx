import { useState } from 'react'
import { Counter as CounterComponent } from '../../components'
import { useAppDispatch, useAppSelector } from '../../hooks'
import {
  decrement,
  increment,
  incrementAsync,
  incrementByAmount,
  incrementIfOdd,
  selectCount,
} from './slice'

/**
 * Counter
 * @returns {React.JSX.Element} Counter component
 */
function Counter(): React.JSX.Element {
  const dispatch = useAppDispatch()
  const count = useAppSelector(selectCount)
  const [incrementAmount, setIncrementAmount] = useState('2')
  const incrementValue = Number(incrementAmount) || 0

  return (
    <CounterComponent
      count={count}
      incrementAmount={incrementAmount}
      onIncrementAmountChange={setIncrementAmount}
      onDecrement={() => dispatch(decrement())}
      onIncrement={() => dispatch(increment())}
      onIncrementAsync={() => dispatch(incrementAsync(incrementValue))}
      onIncrementByAmount={() => dispatch(incrementByAmount(incrementValue))}
      onIncrementIfOdd={() => dispatch(incrementIfOdd(incrementValue))}
    />
  )
}

export default Counter
