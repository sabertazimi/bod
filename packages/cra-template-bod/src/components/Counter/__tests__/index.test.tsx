import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { create } from 'react-test-renderer';
import Counter from '../index';

describe('Counter', () => {
  test('should render correctly (snapshot)', () => {
    const mockClickHandler = jest.fn();

    const tree = create(
      <Counter
        count={0}
        incrementAmount={'2'}
        onIncrementAmountChange={mockClickHandler}
        onDecrement={mockClickHandler}
        onIncrement={mockClickHandler}
        onIncrementAsync={mockClickHandler}
        onIncrementByAmount={mockClickHandler}
        onIncrementIfOdd={mockClickHandler}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Should render accessibility guidelines (AXE)', async () => {
    const mockClickHandler = jest.fn();
    const { container } = render(
      <Counter
        count={0}
        incrementAmount={'2'}
        onIncrementAmountChange={mockClickHandler}
        onDecrement={mockClickHandler}
        onIncrement={mockClickHandler}
        onIncrementAsync={mockClickHandler}
        onIncrementByAmount={mockClickHandler}
        onIncrementIfOdd={mockClickHandler}
      />
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
