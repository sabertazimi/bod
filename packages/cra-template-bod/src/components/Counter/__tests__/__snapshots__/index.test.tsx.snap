// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render correctly (snapshot) 1`] = `
<div>
  <div
    className="row"
  >
    <button
      aria-label="Decrement value"
      className="button"
      onClick={[MockFunction]}
    >
      -
    </button>
    <span
      className="value"
    >
      0
    </span>
    <button
      aria-label="Increment value"
      className="button"
      onClick={[MockFunction]}
    >
      +
    </button>
  </div>
  <div
    className="row"
  >
    <input
      aria-label="Set increment amount"
      className="textbox"
      onChange={[Function]}
      value="2"
    />
    <button
      className="button"
      onClick={[MockFunction]}
    >
      Add Amount
    </button>
    <button
      className="asyncButton"
      onClick={[MockFunction]}
    >
      Add Async
    </button>
    <button
      className="button"
      onClick={[MockFunction]}
    >
      Add If Odd
    </button>
  </div>
</div>
`;
